src/lib/memory/numa_memory_resource.hpp

```diff
@@ -0,0 +1,34 @@
+#pragma once
+#include <jemalloc/jemalloc.h>
+#include <unordered_map>
+
+namespace hyrise {
+
+using ArenaID = uint32_t;
+
+class NumaExtentHooks {
+ public:
+  static void* alloc(extent_hooks_t* extent_hooks, void* new_addr, size_t size, size_t alignment, bool* zero,
+                     bool* commit, unsigned arena_index);
+  static bool dalloc(extent_hooks_t* extent_hooks, void* addr, size_t size, bool committed, unsigned arena_ind);
+
+  static void store_node_id_for_arena(ArenaID, NodeID);
+
+ private:
+  static std::unordered_map<ArenaID, NodeID> node_id_for_arena_id;
+};
```

Comment:
I would refactor this as follows: Replace the class with something like
```cpp
namespace numa_extent_hooks {

void* alloc_hook(extent_hooks_t* extent_hooks, void* new_addr, size_t size, size_t alignment, bool* zero, bool* commit, unsigned arena_index);
bool dalloc_hook(extent_hooks_t* extent_hooks, void* addr, size_t size, bool committed, unsigned arena_ind);

NodeID get_node_id_of_arena(ArenaID);
void store_node_id_for_arena(ArenaID, NodeID);

};
```

where the `get_` and `store_` functions access the global variable `node_id_for_arena_id` only after taking a lock (currently, I think you have a data race in the `NumaMemoryResource` constructor when updating the map). Note that I removed any mention of `node_id_for_arena_id` from the header, because it is actually not needed (noone links to it) and you can just have it in your cpp file (in an anonymous namespace). If you limit access to this global map with the lock, I think it is a fair way to handle it like this (the jemalloc hooks api seems to kind of force you to do so)

Reply:
Note that moving the variable into an anonymous namespace doesn't really hurt debugability, as you can still inspect them in gdb:
```plain
(gdb) info variables niklas
All variables matching regular expression "niklas":

File /home/niklas/Dev/hyrise/src/lib/memory/numa_memory_resource.cpp:
12:     static std::unordered_map<unsigned int, hyrise::NodeID, std::hash<unsigned int>, std::equal_to<unsigned int>, std::allocator<std::pair<unsigned int const, hyrise::NodeID> > > hyrise::(anonymous namespace)::niklas_cool_map;

...
(gdb) p 'hyrise::(anonymous namespace)::niklas_cool_map'
$1 = std::unordered_map with 0 elements
```